parsers/csv/csv.go

```diff
@@ -0,0 +1,152 @@
+package csv
+
+import (
+	"encoding/csv"
+	"errors"
+	"io"
+	"math"
+	"strings"
+	"sync"
+
+	"github.com/Sirupsen/logrus"
+
+	"github.com/honeycombio/honeytail/event"
+	"github.com/honeycombio/honeytail/httime"
+	"github.com/honeycombio/honeytail/parsers"
+)
+
+// Options defines the options relevant to the CSV parser
+type Options struct {
+	Fields          string `long:"fields" description:"Comma separated list of CSV fields, in order."`
+	TimeFieldName   string `long:"timefield" description:"Name of the field that contains a timestamp"`
+	TimeFieldFormat string `long:"time_format" description:"Timestamp format to use (strftime and Golang time.Parse supported)"`
+	NumParsers      int    `hidden:"true" description:"number of csv parsers to spin up"`
+}
+
+// Parser implements the Parser interface
+type Parser struct {
+	conf       Options
+	lineParser parsers.LineParser
+}
+
+// Init constructs our parser from the provided options
+func (p *Parser) Init(options interface{}) error {
+	p.conf = *options.(*Options)
+	if p.conf.Fields == "" {
+		return errors.New("must provide at least 1 field name when parsing CSV lines")
+	}
+	lineParser, err := NewCSVLineParser(p.conf.Fields)
+	if err != nil {
+		return err
+	}
+	p.lineParser = lineParser
+	return nil
+}
+
+type CSVLineParser struct {
+	fields []string
+}
+
+// NewCSVLineParser factory
+func NewCSVLineParser(fieldsString string) (*CSVLineParser, error) {
+	// Is building a reader for every single line a good idea?
+	// Potential for future optimization here
+	reader := strings.NewReader(fieldsString)
+	csvReader := csv.NewReader(reader)
+	csvReader.TrimLeadingSpace = true
+	fields, err := csvReader.Read()
+	if err != nil {
+		logrus.WithError(err).WithField("fields", fieldsString).
+			Error("unable to parse list of fields")
+		return nil, err
+	}
+	logrus.WithFields(logrus.Fields{
+		"fields": fields,
+	}).Debug("generated CSV fields")
+	return &CSVLineParser{fields}, nil
+}
+
+func (p *CSVLineParser) ParseLine(line string) (map[string]interface{}, error) {
+	csvReader := csv.NewReader(strings.NewReader(line))
+	data := make(map[string]interface{})
+	values, err := csvReader.Read()
+	if err != nil && err != io.EOF {
+		logrus.WithError(err).WithField("line", line).
+			Warn("failed to parse line")
+		return nil, err
+	}
+
+	lenValues := len(values)
+	lenFields := len(p.fields)
+
+	if lenValues != lenFields {
```

Comment:
I think it might actually be safer to require the number of fields in each line to match the number of fields in `p.fields`. Otherwise it'd be pretty easy to typo `--csv.fields` and have a silent mismatch between the field names and values in honeycomb. Looks like we can require a specific number of fields by setting `csvReader.FieldsPerRecord`. Thoughts?

Reply:
Yeah I can see that. I was thinking of how we'd want to keep logging (and just warn) even if the number of fields increased or was variable for some reason, but that actually seems like a bad pattern to get into since ordering of fields is important in CSV. I'll change this to error and not send to honeycomb in this scenario.